### PR TITLE
[Core] Allow override env for all task in dag

### PR DESCRIPTION
**Pull request by Michaelvll**
This is required by #3611, as we should be able to propagate the secrets / envs across all the tasks, allowing `sky jobs launch gpt2-pipeline.yaml --env BUCKET_NAME=my-bucket-name`

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/cli.py [hunk 1]
```diff
@@ -765,6 +765,8 @@ def _make_task_or_dag_from_entrypoint_with_overrides(
                     f'WARNING: override params {override_params} are ignored, '
                     'since the yaml file contains multiple tasks.',
                     fg='yellow')
+            for task in dag.tasks:
+                task.update_envs(env)
```
Comment (romilbhardwaj) [inline sky/cli.py:769]: Is this required after removing L93-L98 in `dag_utils.py::load_chain_dag_from_yaml`? 

I may be missing something, but it seems like the `dag = dag_utils.load_chain_dag_from_yaml(entrypoint, env_overrides=env)` call at L758 will now take care of overriding env vars for each task and this should not be required?

**Review by romilbhardwaj — COMMENTED** [review]
thanks @Michaelvll!

**Comment by romilbhardwaj** [comment]
Also I noticed a typo: https://github.com/skypilot-org/skypilot/blob/ea5aa50e19a2c316f2c5dd9bc47245c4b2338b71/llm/gpt-2/README.md?plain=1#L35 it should use `--gpus` instead of `--gpu`. Can we also include that fix in this PR?

File: sky/cli.py (hunk 1 above)
Comment (Michaelvll) [reply sky/cli.py]: Good point! Fixed. Thanks!

File: llm/gpt-2/README.md [hunk 2]
```diff
@@ -30,13 +30,21 @@ sky launch -c gpt2 gpt2.yaml
...
+Or, you can train the model with a single A100, by adding `--gpus A100`:
 ```bash
-sky launch -c gpt2 gpt2.yaml --gpu A100
+sky launch -c gpt2 gpt2.yaml --gpus A100
 ```
 
 ![GPT-2 training with a single A100](https://imgur.com/hN65g4r.png)
 
+
+It is also possible to speed up the training of the model on 8 H100:
```
Comment (romilbhardwaj) [inline llm/gpt-2/README.md:41]:
```suggestion
It is also possible to speed up the training of the model on 8 H100 (2.3x more tok/s than 8x A100s):
```

File: llm/gpt-2/gpt2-pipeline.yaml [hunk 3]
```diff
@@ -0,0 +1,129 @@
+name: gpt2-data
+
+envs:
+  BUCKET_NAME: # Fill in your bucket name
```
Comment (cblmemo) [inline llm/gpt-2/gpt2-pipeline.yaml:4]:
```suggestion
  BUCKET_NAME: # TODO: Fill in your bucket name
```
to align with other examples?

File: llm/gpt-2/gpt2-train.yaml [hunk 4]
```diff
@@ -21,11 +22,11 @@ resources:
 file_mounts:
   ~/.cache/huggingface:
     name: $BUCKET_NAME
+    store: $BUCKET_STORE
     mode: COPY
 
 setup: |
   cd ~
-  pip install tqdm tiktoken requests datasets
```
Comment (romilbhardwaj) [inline llm/gpt-2/gpt2-train.yaml:28]: is this intended?

File: llm/gpt-2/gpt2-train.yaml [hunk 5]
```diff
@@ -1,7 +1,8 @@
-name: train
+name: gpt2-train
 
 envs:
   BUCKET_NAME: # Fill in your bucket name
```
Comment (cblmemo) [inline llm/gpt-2/gpt2-train.yaml:4]:
```suggestion
  BUCKET_NAME: # TODO: Fill in your bucket name
```

**Review by romilbhardwaj — APPROVED** [review]
Thanks @Michaelvll! Ran some manual tests to check envvar behavior, lgtm.

File: sky/utils/dag_utils.py [hunk 6]
```diff
@@ -90,12 +90,6 @@ def load_chain_dag_from_yaml(
         # YAML has only `name: xxx`. Still instantiate a task.
         configs = [{'name': dag_name}]
 
-    if len(configs) > 1:
-        # TODO(zongheng): in a chain DAG of N tasks, cli.py currently makes the
-        # decision to not apply overrides. Here we maintain this behavior. We
-        # can listen to user feedback to change this.
-        env_overrides = None
```
Comment (cblmemo) [inline sky/utils/dag_utils.py:97]: update the docstr of this function as well?

File: sky/cli.py [hunk 7]
```diff
@@ -772,6 +772,7 @@ def _make_task_or_dag_from_entrypoint_with_overrides(
     else:
         task = sky.Task(name='sky-cmd', run=entrypoint)
         task.set_resources({sky.Resources()})
+        task.update_envs(env)
```
Comment (cblmemo) [inline sky/cli.py:775]: could we add a comment saying that for dag yamls, we do env override in the `dag_utils.load_chain_dag_from_yaml` function? I was a little bit confused when I only looked at code updates within this file and saw we changed from updating the env vars for all cases to only updating sky-cmd cases..

File: llm/gpt-2/gpt2-train.yaml (hunk 4 above)
Comment (Michaelvll) [reply llm/gpt-2/gpt2-train.yaml:28]: Yes, it is not needed for training stage, but only for the data processing stage : )

**Review by cblmemo — APPROVED** [review]
Thanks for adding this feature @Michaelvll ! It looks good to me ;) I'm wondering if we need to clarify this semantic changes in the documentation, but we could do it in another PR as well. Maybe left an TODO or file an issue for this if we decided to delay it?

**Pull request merged**